README.md
# SuchBar - Enduser-learnable language for database queries.

[![Crates.io][crates-badge]][crates-url]
[![MIT licensed][mit-badge]][mit-url]
[![Documentation][docs-badge]][docs-url]

[crates-badge]: https://img.shields.io/crates/v/suchbar.svg
[crates-url]: https://crates.io/crates/suchbar
[mit-badge]: https://img.shields.io/badge/license-MIT-blue.svg
[mit-url]: https://github.com/NORICS—NET/suchbar/blob/master/LICENSE
[docs-badge]: https://docs.rs/suchbar/badge.svg
[docs-url]: https://docs.rs/suchbar

A query language that can be learned by normal users, similar to an internet 
search engine.

A query like "`plz=26440-26452 OR (Eisen AND sn!=Hammer*)`" could generate such a SQL:

```sql
SELECT 
    pa.shortname, pa.description, pa.taxnumber, pb.longname, pb.postcode, pb.city, pb.street 
FROM partner_partner pa, partner_branchstore pb 
WHERE 
  pa.id = pb.cmrpartner AND 
  (
    ( pb.postcode>='26440' AND 
      pb.postcode<='26452' 
    ) OR
    ( pa.shortname LIKE '%Eisen%' OR 
      pa.description LIKE '%Eisen%' OR
      pa.taxnumber LIKE '%Eisen%' OR
      pb.city LIKE '%Eisen%' OR
      pb.street LIKE '%Eisen%' 
    ) AND 
    NOT pa.shortname LIKE 'Hammer%' 
 )
LIMIT 20;
```

| Boolean Operator |Alternative Symbol	 | Description                                                                         |
|:----------------:|:------------------:|:---------------------------------------------------------------------------------------|
|       AND        |         &&         | 	Requires both terms on either side of the Boolean operator to be present for a match. |
|       NOT        |         !	         | Requires that the following term not be present.                                       |
|       OR         |    &vert;&vert;    | 	Requires that either term (or both terms) be present for a match.                     |


## Usage: 


### Register the search-fields.

The fields available for searching are registered:
`DbField::new("db-fieldname", db-type, "permission", &["name", "alternative_name", "abbr"]),`   


## Example

```rust 
const SUCHBAR: Suchbar = Suchbar::new(&[
    DbField::new("pa.shortname", TEXT, "STD", &["sname", "sn"]),
    DbField::new("pa.description", TEXT, "STD", &["desc", "d"]),
    DbField::new("pa.taxnumber", VARCHAR(15), "STD", &["ust_id", "tax", "ustid"]),
    DbField::new("pb.city", VARCHAR(35), "STD", &["city"]),
    DbField::new("pb.street", VARCHAR(55), "STD", &["street", "st"]),
    DbField::new("pb.postcode", VARCHAR(5), "STD", &["plz", "zip"]),
]);

fn main() {
    let suche = "plz=26440-26452 OR (Eisen AND sn!=Hammer*)";
    match SUCHBAR.exec(suche) {
        Err(c) => println!("\n{c}"),
        Ok(sr) => {
            let query = format ! (
            "SELECT pa.shortname, pa.description, pa.taxnumber, \
                    pb.longname, pb.postcode AS INTEGER, pb.city, pb.street \
                    FROM partner_partner pa, partner_branchstore pb \
                    WHERE pa.id = pb.cmrpartner{} LIMIT 20",
            sr.to_sql("AND")
            );
        }
    }
}
```/


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## v0.5.1 (2023-12-06)

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 2 commits contributed to the release over the course of 2 calendar days.
 - 223 days passed between releases.
 - 0 commits were understood as [conventional](https://www.conventionalcommits.org).
 - 0 issues like '(#ID)' were seen in commit messages

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **Uncategorized**
    - Updated dependency. (f8c496f)
    - Translated and streamlined. (81a165f)
</details>

## v0.5.0 (2023-04-27)

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 3 commits contributed to the release over the course of 20 calendar days.
 - 34 days passed between releases.
 - 0 commits were understood as [conventional](https://www.conventionalcommits.org).
 - 0 issues like '(#ID)' were seen in commit messages

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **Uncategorized**
    - Parses natural-language-timevalues. (961cdef)
    - Closes #5; test to proof. (3451fbf)
    - Fixes #4: SuchOptions.like_in_numerics added. (59b9f59)
</details>

## v0.4.0 (2023-03-23)

### Other

 - <csr-id-68c12cc2330be55e5ee0e83e5f898dca0f54cdd9/> #2 + #1 - allow iso-dates without quotes.

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 10 commits contributed to the release over the course of 10 calendar days.
 - 1 commit was understood as [conventional](https://www.conventionalcommits.org).
 - 0 issues like '(#ID)' were seen in commit messages

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **Uncategorized**
    - Breaking change: Suchbar is const, exec results a SuchbarResult! (fa0ad14)
    - Fixing !=/NOT-possible error. (6598cbe)
    - Fixes #3. 🤦🏼‍♂ (46aa9dd)
    - Don't use LIKE for non-text fields. (24b5853)
    - #2 + #1 - allow iso-dates without quotes. (68c12cc)
    - WIP - Description extended. (c0c41f6)
    - Allmost useable. (2ea99ec)
    - Refactored to single purpose modules. (627b55d)
    - Proove of concept. (798c913)
    - Initial commit (4a1552d)
</details>



src/error.rs
use crate::error::SuchError::{Denied, LikeNotPossible, ParseError};
use crate::suchbar::Rule;
use std::fmt::{Display, Formatter};

#[derive(Debug)]
pub enum SuchError {
    ParseError(String),
    LikeNotPossible,
    Denied,
}

impl From<pest::error::Error<Rule>> for SuchError {
    fn from(value: pest::error::Error<Rule>) -> Self {
        ParseError(value.to_string())
    }
}

impl Display for SuchError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            ParseError(str) => write!(f, "{str}"),
            LikeNotPossible => write!(f, "LIKE not possible"),
            Denied => write!(f, "DENIED"),
        }
    }
}

impl From<timewarp::TimeWarpError> for SuchError {
    fn from(value: timewarp::TimeWarpError) -> Self {
        ParseError(value.to_string())
    }
}


src/comp_op.rs
use self::CompOp::{Equal, Gt, Gte, Lt, Lte, NotEqual};
use crate::error::SuchError;
use crate::error::SuchError::ParseError;
use std::fmt::{Display, Formatter};
use std::ops::Not;
use std::str::FromStr;

///
#[allow(clippy::upper_case_acronyms)]
#[derive(Copy, Clone, Eq, PartialEq, Debug, Default)]
pub(crate) enum CompOp {
    #[default]
    Equal,
    NotEqual,
    Gt,
    Lt,
    Gte,
    Lte,
}

impl Display for CompOp {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Equal => "=",
                Gt => ">",
                Gte => ">=",
                Lt => "<",
                Lte => "<=",
                NotEqual => "!=",
            }
        )
    }
}

impl FromStr for CompOp {
    type Err = SuchError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "=" | "==" => Ok(Equal),
            ">=" | "=>" => Ok(Gte),
            ">" => Ok(Gt),
            "<=" | "=<" => Ok(Lte),
            "<" => Ok(Lt),
            "!=" | "=!" => Ok(NotEqual),
            _ => Err(ParseError(format!("'{s}' is no comparator!"))),
        }
    }
}

impl Not for CompOp {
    type Output = CompOp;

    fn not(self) -> Self::Output {
        match &self {
            Equal => NotEqual,
            NotEqual => Equal,
            Gt => Lte,
            Gte => Lt,
            Lte => Gt,
            Lt => Gte,
        }
    }
}


src/op_tree.rs
use anyhow::bail;
use std::fmt::{Display, Error, Formatter};
use std::ops::Not;
use std::str::FromStr;

#[derive(Copy, Clone, Eq, PartialEq, Debug)]
pub enum CompOp {
    Equal,
    NotEqual,
    GT,
    LT,
    GTE,
    LTE,
}

impl Default for CompOp {
    fn default() -> Self {
        CompOp::Equal
    }
}

impl Display for CompOp {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        use CompOp::*;
        write!(
            f,
            "{}",
            match self {
                Equal => "=",
                GT => ">",
                GTE => ">=",
                LT => "<",
                LTE => "<=",
                NotEqual => return Err(Error::default()),
            }
        )
    }
}

impl FromStr for CompOp {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "=" => Ok(CompOp::Equal),
            ">=" => Ok(CompOp::GTE),
            ">" => Ok(CompOp::GT),
            "<=" => Ok(CompOp::LTE),
            "<" => Ok(CompOp::LT),
            "!=" => Ok(CompOp::NotEqual),
            _ => Err(()),
        }
    }
}

impl Not for CompOp {
    type Output = CompOp;

    fn not(self) -> Self::Output {
        use CompOp::*;
        match &self {
            Equal => NotEqual,
            NotEqual => Equal,
            GT => LTE,
            GTE => LT,
            LTE => GT,
            LT => GTE,
        }
    }
}

fn try_bool(str: &str) -> anyhow::Result<bool> {
    let str = str.trim().to_ascii_lowercase();
    match str.as_str() {
        "1" | "true" | "wahr" => Ok(true),
        "0" | "false" | "falsch" | "unwahr" => Ok(false),
        _ => bail!("No boolean value: '{str}'"),
    }
}

#[derive(Debug, Copy, Clone)]
pub enum FType {
    VARCHAR(usize),
    INTEGER(u64, u64),
    NUMERIC(u32, u32),
    BOOL,
    DATE,
    TIMESTAMP,
}

#[derive(Debug, Clone)]
pub struct SField {
    pub db_name: &'static str,
    pub db_type: FType,
    pub permission: &'static str,
    pub alias: &'static [&'static str],
}

impl SField {
    pub(crate) const CATCHALL: Self = SField::new("*", FType::VARCHAR(25000), "", &[]);

    /// !important: use lowercase
    pub const fn new(
        db_name: &'static str,
        db_type: FType,
        permission: &'static str,
        alias: &'static [&'static str],
    ) -> Self {
        Self {
            db_name,
            db_type,
            permission,
            alias,
        }
    }

    /// Transforms the given `val` into a EQ-expression. Replaces symbols into a sql-save query.
    pub(crate) fn try_sql_eq(&self, eq: CompOp, val: &str) -> anyhow::Result<String> {
        use FType::*;
        let Self {
            db_name, db_type, ..
        } = self;
        match db_type {
            VARCHAR(a) if *a > val.len() => {
                Ok(format!("{db_name}{eq}'{}'", val.replace('\'', "''")))
            }
            BOOL if eq == CompOp::Equal => Ok(format!(
                "{}{db_name}",
                if try_bool(val)? { "" } else { "NOT " }
            )),
            DATE | TIMESTAMP => Ok(format!("{db_name}{eq}'{}'", val.replace('\'', "''"))),
            INTEGER(min, max) => {
                let cval = val.replace(',', ".");
                match u64::from_str(&cval) {
                    Ok(d) if d <= *max && d >= *min => Ok(format!("{db_name}{eq}{cval}")),
                    _ => {
                        //  println!("No Integer value '{val}'");
                        bail!("No Integer value '{val}'");
                    }
                }
            }
            NUMERIC(len, _) => {
                let cval = val.replace(',', ".");
                match f64::from_str(&cval) {
                    Ok(_) if cval.len() < (*len + 1) as usize => Ok(format!("{db_name}{eq}{cval}")),
                    _ => {
                        //  println!("No numeric value '{val}'");
                        bail!("No numeric value '{val}'");
                    }
                }
            }
            _ => bail!("Don't know how to handle: {db_type:?} = '{val}'"),
        }
    }

    /// Transforms the given `val` into a LIKE-expression. Replaces key-symbols from glob-style to
    /// form a sql-save query.
    pub(crate) fn try_sql_like(&self, val: &str) -> anyhow::Result<String> {
        let Self {
            db_name, db_type, ..
        } = self;
        let escaper = |c: char| match c {
            '?' => String::from("_"),
            '*' => String::from("%"),
            '\'' => String::from("''"),
            '_' | '%' => format!("\\{c}"),
            _ => String::from(c),
        };
        match db_type {
            FType::VARCHAR(a) if val.len() <= *a => Ok(format!(
                "{db_name} IS LIKE '%{}%'",
                val.chars().map(escaper).collect::<String>()
            )),
            FType::DATE if val.len() <= 10 => Ok(format!(
                "{db_name} IS LIKE '%{}%'",
                // pre-checking dates is to cumbersome, let the db make its job.
                val.chars().map(escaper).collect::<String>()
            )),

            _ => bail!("{db_name} ({db_type:?})'{val}' is not compatible "),
        }
    }
}

#[cfg(test)]
mod should {

    use crate::op_tree::FType::{BOOL, DATE, VARCHAR};
    use crate::op_tree::{CompOp, FType, SField};
    use crate::sql_term::SQLTerm::*;

    const ARTIKEL: SField = SField::new(
        "article",
        VARCHAR(200),
        "READ_OFFER",
        &["artnr", "artikelnr"],
    );
    const ACTIVE: SField = SField::new("aktiv", BOOL, "READ_OFFER", &["akt"]);
    const END_DATE: SField = SField::new("end_date", DATE, "READ_OFFER", &["enddate", "end_date"]);
    const NAME: SField = SField::new("ma_active", VARCHAR(32), "READ_OFFER", &["akt"]);
    const PRICE: SField = SField::new(
        "price",
        FType::INTEGER(0, 2000),
        "READ_OFFER_PRICE",
        &["price"],
    );

    #[test]
    fn op_to_sql() {
        let df = AND(vec![
            VALUE(ARTIKEL, CompOp::GT, "1245667".into()),
            OR(vec![
                NOT(Box::new(VALUE(ACTIVE, CompOp::Equal, "false".into()))),
                VALUE(END_DATE, CompOp::LTE, "2022-12-24".into()),
                LIKE(NAME, "Micha's cat".into()),
            ]),
        ]);
        assert_eq!(
            df.to_sql().unwrap(),
            "( article>'1245667' AND ( NOT ( NOT aktiv ) OR end_date<=\
            '2022-12-24' OR ma_active IS LIKE '%Micha''s cat%' ) )"
        );

        let df = VALUE(PRICE, CompOp::Equal, "1000.0".into());
        assert_eq!(df.to_sql().unwrap_or_default(), "");
        let df = VALUE(PRICE, CompOp::Equal, "1000".into());
        assert_eq!(df.to_sql().unwrap_or_default(), "price=1000");
    }
}


src/db_field.rs
use self::DbType::{BOOL, DATE, INTEGER, NUMERIC, TEXT, TIMESTAMP, VARCHAR};
use crate::comp_op::CompOp;
use crate::error::SuchError;
use crate::error::SuchError::ParseError;
use timewarp::{date_matcher, Direction, Doy};

fn try_bool(str: &str) -> Result<bool, SuchError> {
    let str = str.trim().to_ascii_lowercase();
    match str.as_str() {
        "1" | "true" | "wahr" | "yes" | "ja" | "y" | "j" | "t" | "w" => Ok(true),
        "0" | "false" | "falsch" | "unwahr" | "no" | "not" | "nein" | "n" | "f" => Ok(false),
        _ => Err(ParseError(format!("No boolean value: '{str}'"))),
    }
}

fn timestamp_checker(str: String) -> Result<String, SuchError> {
    if str.chars().any(|a| match a {
        '-' | ':' | ' ' | '%' => false,
        _ => !a.is_ascii_digit(),
    }) {
        Err(ParseError(String::from("No date")))
    } else if str.len() == 10 {
        Ok(str + " 00:00:00")
    } else {
        Ok(str)
    }
}

/// Definition of a Database-Field.   
#[derive(Debug, Clone)]
pub struct DbField {
    pub db_name: &'static str,
    pub db_type: DbType,
    pub permission: &'static str,
    pub alias: &'static [&'static str],
}

impl DbField {
    /// !important: use lowercase
    #[must_use]
    pub const fn new(
        db_name: &'static str,
        db_type: DbType,
        permission: &'static str,
        alias: &'static [&'static str],
    ) -> Self {
        Self {
            db_name,
            db_type,
            permission,
            alias,
        }
    }

    /// Transforms the given `val` into a EQ-expression. Replaces symbols into a sql-save query.
    ///
    /// # Errors
    /// May fail if `val` can't be parsed to the needed type.
    pub(crate) fn try_sql_eq(
        &self,
        eq: CompOp,
        val: &str,
        d: Direction,
    ) -> Result<String, SuchError> {
        let Self {
            db_name, db_type, ..
        } = self;

        match db_type {
            BOOL => {
                let not = try_bool(val)? == (eq == CompOp::Equal);
                Ok(format!("{db_name}{}", if not { "" } else { "=false" }))
            }
            NUMERIC(_, _) | INTEGER(_, _) => Ok(format!("{db_name}{eq}{}", db_type.sql_safe(val)?)),
            DATE => {
                let date = date_matcher(Doy::today(), d, val).map(|d| d.start())?;
                Ok(format!("{db_name}{eq}'{date:#}'"))
            }
            _ => Ok(format!("{db_name}{eq}'{}'", db_type.sql_safe(val)?)),
        }
    }

    /// Transforms the given `val` into a LIKE-expression. Replaces key-symbols from glob-style to
    /// form a sql-save query.
    pub(crate) fn try_sql_like(&self, val: &str) -> Result<String, SuchError> {
        let Self {
            db_name, db_type, ..
        } = self;
        match db_type {
            VARCHAR(_) | TEXT => Ok(format!("{db_name} LIKE '{}'", db_type.sql_safe(val)?)),
            DATE | TIMESTAMP => Err(SuchError::LikeNotPossible),
            _ => Ok(format!("{db_name}::TEXT LIKE '{}'", db_type.sql_safe(val)?)),
        }
    }

    #[must_use]
    pub fn is_text(&self) -> bool {
        matches!(self.db_type, TEXT | VARCHAR(_))
    }

    /// Returns all aliases by which this `DbField` can be used.
    #[must_use]
    pub fn aliases(&self) -> String {
        format!("[{}]", self.alias.join(", "))
    }

    /// Returns a simplified Type.
    #[must_use]
    pub fn db_type(&self) -> String {
        self.db_type.name()
    }
}

/// Type of the field in the database.
/// It is used to decide in which columns valid hits could be found and how
/// the query must be set up.
#[derive(Debug, Copy, Clone)]
pub enum DbType {
    /// Text in a max. length.
    VARCHAR(usize),
    /// Text in any length.
    TEXT,
    /// The min / max-value this INTEGER is valid.
    INTEGER(u64, u64),
    /// The precision of a numeric is the total count of significant digits
    /// in the whole number, that is, the number of digits to both sides of
    /// the decimal point. The scale of a numeric is the count of decimal
    /// digits in the fractional part, to the right of the decimal point.
    /// So the number 23.5141 has a precision of 6 and a scale of 4.
    NUMERIC(u32, u32),
    /// Interprets:
    /// * True: `1`, `true`, `wahr`, `yes`, `ja`, `y`, `j`, `t`, `w`
    /// * False: `0`, `false`, `falsch`, `unwahr`, `no`, `not`, `nein`, `n`, `f`
    BOOL,
    /// Date without time
    DATE,
    /// Date with time
    TIMESTAMP,
}

impl DbType {
    fn sql_safe(&self, val: &str) -> Result<String, SuchError> {
        let escaper = |c: char| match c {
            '?' => String::from("_"),
            '*' => String::from("%"),
            '\'' => String::from("''"),
            '_' | '%' => format!("\\{c}"),
            _ => String::from(c),
        };
        self.checker(val.chars().map(escaper).collect::<String>())
    }

    fn checker(&self, val: String) -> Result<String, SuchError> {
        use std::str::FromStr;
        match self {
            VARCHAR(a) if val.len() > *a => Err(ParseError(format!("Value: '{val}' to long"))),
            VARCHAR(_) | TEXT => Ok(val),
            TIMESTAMP => timestamp_checker(val),
            INTEGER(min, max) => {
                let c_val = val.replace(',', ".");
                match u64::from_str(&c_val.replace('%', "")) {
                    Ok(d) if d <= *max && d >= *min => Ok(c_val),
                    _ => Err(ParseError(format!("No Integer value '{val}'"))),
                }
            }
            NUMERIC(len, _) => {
                let c_val = val.replace(',', ".");
                let number = c_val.replace('%', "");
                match f64::from_str(&number) {
                    Ok(_) if number.len() < (len + 1) as usize => Ok(c_val),
                    _ => Err(ParseError(format!("No Numeric value '{val}'"))),
                }
            }
            _ => Err(ParseError(format!(
                "Don't know how to handle: {self:?} = '{val}'"
            ))),
        }
    }

    #[must_use]
    pub fn name(&self) -> String {
        match self {
            VARCHAR(_) | TEXT => "TEXT",
            INTEGER(_, _) | NUMERIC(_, _) => "NUMBER",
            BOOL => "BOOL",
            DATE | TIMESTAMP => "TIME",
        }
        .into()
    }
}

#[derive(Debug)]
pub struct SortField {
    pub desc: bool,
    pub field: DbField,
}

impl SortField {
    pub fn to_sql(&self) -> String {
        format!(
            "{}{}",
            self.field.db_name,
            if self.desc { " DESC" } else { "" }
        )
    }
}

#[cfg(test)]
mod should {
    use crate::comp_op::CompOp;
    use crate::db_field::DbField;
    use crate::db_field::DbType::{BOOL, DATE, INTEGER, VARCHAR};
    use crate::sql_term::SQLTerm::{AND, LIKE, NOT, OR, VALUE};
    use crate::DbType::TIMESTAMP;
    use timewarp::Direction::From;

    const ARTIKEL: DbField = DbField::new(
        "article",
        VARCHAR(200),
        "READ_OFFER",
        &["artnr", "artikelnr"],
    );
    const ACTIVE: DbField = DbField::new("aktiv", BOOL, "READ_OFFER", &["akt"]);
    const END_DATE: DbField =
        DbField::new("end_date", DATE, "READ_OFFER", &["enddate", "end_date"]);
    const NAME: DbField = DbField::new("ma_active", VARCHAR(32), "READ_OFFER", &["akt"]);
    const PRICE: DbField = DbField::new("price", INTEGER(0, 2000), "READ_OFFER_PRICE", &["price"]);
    const CHANGED: DbField =
        DbField::new("changed", TIMESTAMP, "READ_OFFER", &["changed", "updated"]);

    #[test]
    fn op_to_sql() {
        let df = AND(vec![
            VALUE(ARTIKEL, CompOp::Gt, From, "1245667".into()),
            OR(vec![
                NOT(Box::new(VALUE(ACTIVE, CompOp::Equal, From, "false".into()))),
                VALUE(END_DATE, CompOp::Lte, From, "2022-12-24".into()),
                LIKE(NAME, "Micha's cat*".into()),
            ]),
        ]);
        assert_eq!(
            df.to_sql().unwrap(),
            "( article>'1245667' AND ( NOT aktiv=false OR end_date<=\
            '2022-12-24' OR ma_active LIKE 'Micha''s cat%' ) )"
        );

        let df = VALUE(PRICE, CompOp::Equal, From, "1000.0".into());
        assert!(df.to_sql().is_err());
        let df = VALUE(PRICE, CompOp::Equal, From, "1000".into());
        assert_eq!(df.to_sql().unwrap_or_default(), "price=1000");

        let df = VALUE(CHANGED, CompOp::Gte, From, "2022-09-01".into());
        assert_eq!(
            df.to_sql().unwrap_or_default(),
            "changed>='2022-09-01 00:00:00'"
        );
        let df = VALUE(CHANGED, CompOp::Lt, From, "2022-09-01 23:30:00".into());
        assert_eq!(
            df.to_sql().unwrap_or_default(),
            "changed<'2022-09-01 23:30:00'"
        );

        let df = VALUE(CHANGED, CompOp::Lt, From, "2022-09-01 23:30:00 MEZ".into());
        assert!(df.to_sql().is_err());
    }
}


src/sql_term.rs
use crate::comp_op::CompOp;
use crate::db_field::DbField;
use crate::error::SuchError;
use crate::error::SuchError::ParseError;
use std::fmt::{Display, Formatter};
use std::ops::Deref;
use timewarp::Direction;

#[allow(clippy::upper_case_acronyms)]
#[derive(Debug)]
pub(crate) enum SQLTerm {
    AND(Vec<SQLTerm>),
    OR(Vec<SQLTerm>),
    NOT(Box<SQLTerm>),
    VALUE(DbField, CompOp, Direction, String),
    LIKE(DbField, String),
    DENIED,
}

impl SQLTerm {
    pub fn to_sql(&self) -> Result<String, SuchError> {
        use SQLTerm::{AND, DENIED, LIKE, NOT, OR, VALUE};
        match self {
            OR(vec) => explode(vec, " OR "),
            AND(vec) => explode(vec, " AND "),
            NOT(val) => match val.deref() {
                // NOT( NOT(val)) => val
                NOT(inner) => inner.to_sql(),
                _ => Ok(format!("NOT {}", val.to_sql()?)),
            },
            VALUE(f, eq, d, v) => val_sql(f, *eq, v, *d),
            LIKE(f, v) => f.try_sql_like(v),
            DENIED => Err(SuchError::Denied),
        }
    }
}

fn val_sql(f: &DbField, eq: CompOp, v: &str, d: Direction) -> Result<String, SuchError> {
    if v.contains('*') {
        f.try_sql_like(v)
    } else {
        f.try_sql_eq(eq, v, d)
    }
}

fn explode(vec: &[SQLTerm], sep: &str) -> Result<String, SuchError> {
    let v = vec
        .iter()
        .filter_map(|op| op.to_sql().ok())
        .collect::<Vec<String>>();
    match v.len() {
        0 => Err(ParseError("Empty SQLTerm!".to_string())),
        1 => Ok(v[0].clone()),
        _ => Ok(format!("( {} )", v.join(sep))),
    }
}

impl Default for SQLTerm {
    fn default() -> Self {
        SQLTerm::OR(vec![])
    }
}

impl Display for SQLTerm {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.to_sql().unwrap_or_default())
    }
}


src/lib.rs
//! # Enduser-learnable language for database queries.
//!
//! A query language that can be learned by normal users, similar to an internet
//! search engine. To empower users to filter and sort data in a kind of natural
//! language while preserving permissions.     
//!
//! The search query entered by the user is parsed and converted into an SQL WHERE
//! clause. For the best possible user experience, field-free queries are sent to
//! all syntactically eligible fields as similarity (LIKE) searches.   
//!
//! ## Example
//!
//! A query like `"plz=26440-26452 OR (Eisen AND sn!=Hammer*); plz"` could generate a
//! WHERE-Clause for such kind of SELECT.
//!
//! ```sql
//! SELECT
//!     pa.shortname, pa.description, pa.taxnumber, pb.longname, pb.postcode, pb.city, pb.street
//! FROM
//!   partner_partner pa, partner_branchstore pb
//! WHERE
//!   pa.id = pb.cmrpartner AND
//!   (
//!     ( pb.postcode>='26440' AND
//!       pb.postcode<='26452'
//!     ) OR
//!     ( pa.shortname LIKE '%Eisen%' OR
//!       pa.description LIKE '%Eisen%' OR
//!       pa.taxnumber LIKE '%Eisen%' OR
//!       pb.city LIKE '%Eisen%' OR
//!       pb.street LIKE '%Eisen%'
//!     ) AND
//!     NOT pa.shortname LIKE 'Hammer%'
//!   )
//! ORDER BY pb.postcode
//! ;
//! ```
//!
//! ## Usage
//! ```rust
//! use permeable::AllowAllPermission;
//! use suchbar::{DbField, Suchbar};
//! use suchbar::DbType::*;
//!
//! const SUCHBAR: Suchbar = Suchbar::new(&[
//!     //            fieldname in SQL, fieldtype    Permission  Aliases for query
//!     DbField::new("pa.shortname",    TEXT,        "STD",      &["sname", "sn"]),
//!     DbField::new("pa.description",  TEXT,        "STD",      &["desc", "d"]),
//!     DbField::new("pa.taxnumber",    VARCHAR(15), "STD",      &["ust_id", "tax", "ustid"]),
//!     DbField::new("pb.city",         VARCHAR(35), "STD",      &["city", "ort"]),
//!     DbField::new("pb.street",       VARCHAR(55), "STD",      &["street", "st"]),
//!     DbField::new("pb.postcode",     VARCHAR(5),  "STD",      &["plz", "zip"]),
//! ]);
//!
//! fn main() {
//!     let suche = "plz=26440-26452 OR (Eisen AND sn!=Hammer*); plz";
//!     match SUCHBAR.exec(&AllowAllPermission(), suche) {
//!         Err(error) => println!("\n{error}"),
//!         Ok(sr) => {
//!             let query = format ! (
//!             "SELECT pa.shortname, pa.description, pa.taxnumber, \
//!                     pb.longname, pb.postcode AS INTEGER, pb.city, pb.street \
//!                     FROM partner_partner pa, partner_branchstore pb \
//!                     WHERE pa.id = pb.cmrpartner{} LIMIT 20",
//!             sr.to_sql("AND")
//!             );
//!         }
//!     }
//! }
//! ```

mod comp_op;
mod db_field;
mod error;
mod sql_term;
mod suchbar;

#[macro_use]
extern crate pest_derive;

pub use crate::db_field::{DbField, DbType};
pub use crate::suchbar::{SuchOptions, Suchbar, WhereClause};
